compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -2799,19 +2800,39 @@ object Types {
     }
   }
 
-  object MethodType extends MethodTypeCompanion {
-    def apply(paramNames: List[TermName])(paramInfosExp: MethodType => List[Type], resultTypeExp: MethodType => Type)(implicit ctx: Context): MethodType =
-      checkValid(unique(new CachedMethodType(paramNames)(paramInfosExp, resultTypeExp)))
-  }
+  object MethodKinds {
+    type MethodKind = Byte
 
-  object JavaMethodType extends MethodTypeCompanion {
-    def apply(paramNames: List[TermName])(paramInfosExp: MethodType => List[Type], resultTypeExp: MethodType => Type)(implicit ctx: Context): MethodType =
-      unique(new JavaMethodType(paramNames)(paramInfosExp, resultTypeExp))
+    val Plain: MethodKind = 0: Byte
+
+    // Flags
+    val JavaKind: MethodKind = 1: Byte
+    val ImplicitKind: MethodKind = 2: Byte
+
+    def makeMethodKind(isJava: Boolean, isImplicit: Boolean): MethodKind = {
+      var mk: Int = Plain
+      if (isJava) mk |= JavaKind
+      if (isImplicit) mk |= ImplicitKind
+      mk.toByte
+    }
+
+    implicit class MethodKindsOps(mk: MethodKind) {
+      def is(that: MethodKind): Boolean = (mk & that) == that
+    }
   }
 
-  object ImplicitMethodType extends MethodTypeCompanion {
-    def apply(paramNames: List[TermName])(paramInfosExp: MethodType => List[Type], resultTypeExp: MethodType => Type)(implicit ctx: Context): MethodType =
-      checkValid(unique(new ImplicitMethodType(paramNames)(paramInfosExp, resultTypeExp)))
+  object MethodType extends MethodTypeCompanion { self =>
+    import MethodKinds._
+
```

Comment:
Since MethodTypes are created very often (~500K for Dotty bootstrap) and there are only a small number of different ones, I would not do the `withKind` scaffolding. I would create a finite number of companions instead and use the kind as a parameter.

Actually,  I think a simpler scheme is to forgo kinds altogether and only use companions.
Yes, it will get a little bit more complicated once we add `unused`, but it's still much easier than defining separate kinds.

Reply:
That design is a lot simpler. The only drawback is that we will have an additional indirection to know is a type is a Java method type or implicit method type as that knowlage is in the companion.